Implementing circle

thomasguillard opened the pull request:
This is adding a new method to plot circle on the screen.
It is called using screen.circle() and requires a position. You can also give it a radius and a color.

This will let users have access to a new shape, ideal for projects using balls, planets, coins (and everything round really)

thomasguillard committed:
Adding screen.circle() to graphics.py

This new method lets you draw circles.
It needs a position but defaults radius, color and alignment.
I basically copied screen.rectangle and adapted it using pygame methods.

thomasguillard committed:
Simplify things, no need for alignment yet

thomasguillard committed:
Validation for radius as an int only

thomasguillard committed:
Added a : for an if statement

thomasguillard committed:
Realised sel._fill is basically creating a rectangle.
So I directly used pygame.draw.circle

thomasguillard committed:
Data Validation for xy arg

joerick commented:
Hi @thomasguillard - this is great. Thank you.

I've been thinking about the API for this... imagine we also have an `oval()` function (we will one day). What will be API for that be?

For consistency’s sake, I think parameters should match the rectangle function, that is

```
def oval(self, xy=None, size=(100, 100), color='grey', align='center'):
```

So I think we should make sure that the circle function is consistent with this function. I note that the above code (where width and height are equal) would actually draw a circle! So how about this for the circle API:

```
def circle(self, xy=None, size=100, color='grey', align='center')
```

then we have the advantage that the APIs are all consistent.

An added advantage of this is that the circle function can call the oval function. I’d actually argue that they should be the same function, just with a different name/documentation - and `oval()` can take a number or a 2-tuple - there’s no point in the library being pedantic.

What do you think?

thomasguillard commented:
Hi @joerick !
This is a very good point, I didn't think of that.
This is also making sense for being consistent with another one we will consider later I'm sure which is the arc shape.
Pygame doc says: arc(Surface, color, **Rect**, start_angle, stop_angle, width=1)
This is also based on using a rectangle to draw a specific ellipse section so that should also follow this consistency.

I really like your idea, the circle function should definitively just be a special case of the oval function.

I'll work on that!

I can also see that you want to get the align argument in this function too. I left it aside and actually forgot to include it at the end, I will look into it as well.

Thanks for your comment and for sharing your opinion!

thomasguillard committed:
Added an oval method.
Linked circle to the oval method, as it is just a special case.

thomasguillard committed:
Correcting how to call oval from circle

thomasguillard committed:
Default size of circle is now 100.
Documentation and reporting is consistent with new naming convention.
Comment explaining why we call oval in circle.

thomasguillard committed:
Documentation and comments

joerick commented on tingbot/graphics.py:
```diff
@@ -217,6 +217,51 @@ def text(self, string, xy=None, color='grey', align='center', font=None, font_si
+                if `xy` is None. Defaults to 'center'.
+        """
+        if xy is None:
+            raise ValueError('no value given for xy')
```
`xy = None` is a valid parameter - it uses `align` to find positioning

joerick commented on tingbot/graphics.py:
```diff
@@ -217,6 +217,51 @@ def text(self, string, xy=None, color='grey', align='center', font=None, font_si
+        elif type(xy) is not tuple:
+            raise ValueError('xy should be a 2-tuple')
+        elif len(xy) != 2:
+            raise ValueError('xy should be a 2-tuple')
```
These xy checks will fail when None is passed, so they should be wrapped in `if xy is not None:`

They should also be moved to the oval function, as this circle function doesn't need the check parameters - just pass on to oval.

furbrain replied:
I'd argue that you shouldn't be testing directly for whether xy is a tuple or not - what if a `list` was passed? or a `NamedTuple`?
My preferred solution would be to not even test this at all - if xy allows xy[0] and xy[1], then it will just work, or you will get an exception raised later on.
If you really feel the need to check the type may I suggest:
`
elif not isinstance(xy,collections.Sequence):
`
This will allow through anything based on collections.Sequence, which includes lists, tuples, NamedTuples etc.

joerick replied:
Thanks @furbrain, I've had a think and I'd agree. Really the purpose of this type checking is to present more helpful error messages to the user, so we shouldn't raise an exception on anything that would otherwise work.

It's more idiomatic to just let Python raise an exception, but when beginners are presented with an error like `TypeError: 'int' object has no attribute '__getitem__'` there's nothing there that's going to help them debug the problem (and googling it isn't going to help)

How about this for the type check:

```
if not (hasattr(xy, '__len__')
        and len(xy) >= 2
        and isinstance(xy[0], numbers.Real)
        and isinstance(xy[1], numbers.Real)):
    raise TypeError('xy should contain two numbers')
```

---

side note:  
I'm not a big fan of 'type' checking overall, since as a library author you're asserting yourself over the user (assuming they don't know what they're doing). However we need to improve the error messages in tingbot-python to improve accessibility. I think in Python it's possible to do this without relying on `isinstance` (except with numbers, numbers are numbers!).

thomasguillard replied:
Hello!
So should we do some checking on types using your suggestion @joerick or should i leave it aside and rely on python exceptions?

This question is basically impacting all of the changes you are requesting I think.

joerick replied:
Sorry, I wasn't clear - the code I posted above is what I'd consider acceptable parameter checking, since it doesn't use isinstance to check a type hierarchy, it only checks what an object can actually _do_ (this is called duck typing). So I'd like to keep some parameter checking using the code above.

thomasguillard replied:
So...
`if not (hasattr(xy, '__len__')
        and len(xy) >= 2
        and isinstance(xy[0], numbers.Real)
        and isinstance(xy[1], numbers.Real)):
    raise TypeError('xy should contain two numbers')`
is not needed?

thomasguillard replied:
Ho... I think I get it, it is still using isinstance but not for its type, but what it can do! so it should be in!
Right?

joerick replied:
yep! :)

joerick commented on tingbot/graphics.py:
```diff
@@ -217,6 +217,51 @@ def text(self, string, xy=None, color='grey', align='center', font=None, font_si
+        """
+        if len(size) != 2:
+            raise ValueError('size should be a 2-tuple')
+
```
To give nice error messages, we should check that the elements of `size` are numbers. But don't check using `type(el) is not int` - that would stop floating point numbers from getting through. Instead use `isinstance(el, numbers.Real)`.

joerick commented on tingbot/graphics.py:
```diff
@@ -217,6 +217,51 @@ def text(self, string, xy=None, color='grey', align='center', font=None, font_si
+            raise ValueError('xy should be a 2-tuple')
+
+        if type(size) is not int:
+            raise ValueError('size should be an integer')
```
Again, allow floats using `isinstance(el, numbers.Real)`

thomasguillard replied:
I suppose I shouldn't do the isinstance() checks, considering what you said before?

joerick reviewed (changes requested):
Nearly there! This is looking great @thomasguillard :)

thomasguillard committed:
Moved xy checks to Oval and modified check types.
It now allows passing more elements than initially required and floats are allowed too as long as everything is valid.

thomasguillard committed:
Added lot of checks for validation xy and size.

thomasguillard commented:
Hi @joerick ,
I have been adding a bunch of checks to xy and size but another idea came to my mind doing so.

If we want some consistency for the API calls, wouldn't it be better to perform checks for xy and size in all the screen methods?

If so, it would be more efficient to create checking functions for xy and size separatly and calls them in each screen functions.

What do you think?

joerick commented:
Hey thomasguillard, sorry for the delay in getting back to you. I think there is a bigger job improving parameter checking throughout the library but let's stay focused for this PR and just worry about circle. 

Just a few notes on style 
1. could you try and write the code so that parameter checking comes at the start of the methods and isn't getting muddled into the actual drawing code? As a rule of thumb, try to only have one line that's doing the actual 'work' of the method and have that at the end
2. A lot of the comments in this code are redundant- you're just saying what each line does, but that should be obvious by reading the code.  Could they be stripped out unless they're saying something that the code isn't saying?

thomasguillard commented:
Hi @joerick !

I'll look into it! ;-)

thomasguillard commented:
All (daft) comments have been taken off. Don't know what I was thinking about...

I don't get your request about having one line doing the drawing, since we introduced checking I need two different cases for each method.

One if you send a number to circle, and another if you send a list to grab the first element.
One if you send a 2-tuple to oval, and another if you send a list to grab the two first elements?

Or am I missing something?

joerick commented on tingbot/graphics.py:
```diff
@@ -217,6 +217,77 @@ def text(self, string, xy=None, color='grey', align='center', font=None, font_si
+                 if len(xy) >= 2:
+                    # as long as the first two are valid numbers.
+                    if not(isinstance(xy[0], numbers.Real) and isinstance(xy[1], numbers.Real)):
+                        raise TypeError('xy should contain two numbers')
```
Let's look at the type checks again - your new versions aren't trapping all the error cases

For example, in the lines above, if xy is `(1,)` (a tuple with only one element), it'll pass through this code without raising an exception. Compare to the code I posted above-

```
if not (hasattr(xy, '__len__')
        and len(xy) >= 2
        and isinstance(xy[0], numbers.Real)
        and isinstance(xy[1], numbers.Real)):
    raise TypeError('xy should contain two numbers')
```

In this, if any of the checks fail the exception is raised. (Btw, I'd quite like the checks in the code to look like the above!)

joerick commented on tingbot/graphics.py:
```diff
@@ -217,6 +217,77 @@ def text(self, string, xy=None, color='grey', align='center', font=None, font_si
+                    pygame.draw.ellipse(self.surface, _color(color), pygame.Rect(xy, size[0:2]), 0) 
+            else:
+                xy = _topleft_from_aligned_xy(xy, align, size, self.size)
+                pygame.draw.ellipse(self.surface, _color(color), pygame.Rect(xy, size), 0) 
```
instead of checking that size is greater than 2 and having the size[0:2] stuff, i'd just raise an error if the user passed in too many elements (there's basically no reason a user would do this on purpose). that simplifies this code a bit.

joerick commented on tingbot/graphics.py:
```diff
@@ -217,6 +217,77 @@ def text(self, string, xy=None, color='grey', align='center', font=None, font_si
+                        self.oval(xy,(size[0],size[0]), color, align)
+            else:
+                # a circle is just a special case of an oval shape:
+                self.oval(xy, (size,size), color, align)
```
here, you can simplify the code by reassigning to the size variable to avoid the repeating self.oval line. So we can change the above code to:

```
if isinstance(size, numbers.Real):
    size = (size, size)
elif hasattr(size, '__len__') and len(size) == 2:
    pass
else:
    raise TypeError('size should be a number or a 2-tuple')

self.oval(xy=xy, size=size, color=color, align=align)
```
